access/ms-teams/msapi/bot_framework_client.go

```diff
@@ -0,0 +1,135 @@
+package msapi
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	botFrameworkAuthScope     = "https://api.botframework.com/.default"
+	botFrameworkBaseURL       = "https://smba.trafficmanager.net"
+	botFrameworkDefaultRegion = "uk"
+	botFrameworkVersion       = "v3"
+)
+
+// BotFrameworkClient represents client to MS Graph API
+type BotFrameworkClient struct {
+	Client
+}
+
+// ChatMember represents member of a chat
+type ChatMember struct {
+	ID                string `json:"id"`
+	ObjectID          string `json:"objectId"`
+	Name              string `json:"name"`
+	Email             string `json:"email"`
+	UserPrincipalName string `json:"userPrincipalName"`
+	TenantID          string `json:"tenantId"`
+	UserRole          string `json:"userRole"`
+}
+
+// PostActivityResponse represents json response with a single id field
+type PostActivityResponse struct {
+	ID string `json:"id"`
+}
+
+// botError represents MS Graph error
+type botError struct {
+	E struct {
+		Code    string `json:"code"`
+		Message string `json:"message"`
+	} `json:"error"`
+}
+
+// Error returns error string
+func (e *botError) Error() string {
+	return e.E.Code + " " + e.E.Message
+}
+
+// NewBotFrameworkClient creates MS Graph API client
+func NewBotFrameworkClient(config Config) *BotFrameworkClient {
+	region := config.Region
+	if region == "" {
+		region = botFrameworkDefaultRegion
+	}
+
+	baseURL := config.url.botFrameworkBaseURL
+	if baseURL == "" {
+		baseURL = botFrameworkBaseURL
+	}
+
+	baseURL = baseURL + "/" + region + "/" + botFrameworkVersion
+
+	return &BotFrameworkClient{
+		Client: Client{
+			token:   tokenWithTTL{scope: botFrameworkAuthScope, baseURL: config.url.tokenBaseURL},
+			baseURL: baseURL,
+			config:  config,
+		},
+	}
+}
+
+// GetChatMembers returns members of a chat
+func (c *BotFrameworkClient) GetChatMembers(ctx context.Context, chatID string) ([]ChatMember, error) {
+	var m []ChatMember
+
+	request := request{
+		Method:   http.MethodGet,
+		Path:     "conversations/" + chatID + "/members",
+		Response: &m,
+		Err:      &botError{},
+	}
+
+	err := c.request(ctx, request)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return m, nil
+}
+
+// PostAdaptiveCardActivity sends an activity to the chat, content is AdaptiveCard
+func (c *BotFrameworkClient) PostAdaptiveCardActivity(ctx context.Context, botID, chatID, card, updateID string) (string, error) {
+	body := `
+		{
+			"type": "message",
+			"from": {
+				"id": "29:` + botID + `",
+				"name": "TeleBot"
```

Comment:
What's "29"? And where does this "name" show up?

Reply:
29 is an internal API designator used to determine a resource type (bot in our case). Name does not show up anywhere, it is used as a placeholder here. The problem is the following. Normally, MS bot is supposed to receive events from the users. When a user (or an admin) installs a bot for a user, the "StartConversation" event is generated. This event contains "29:<id>" along with some other information required to continue the conversation from bot's side. Bot is supposed to process this event and save all the information in the persistent storage for later usage.That's the normal way of how MS bots work. In our case there's no answering part, so I have to improvise. This approach is confirmed by an article from one of MS Teams engineers, where he proposes to do the same in case an original event got lost or a bot is totally passive and can't process any evetns.